File: Sudoku_From_Image/Sudoku_From_Image.jl
using Images, ImageView, Tesseract, ImageFiltering

# Load image and preprocess
img = load("Sudoku_From_Image/sudoku.jpg")
img_gray = Gray.(img)
img_pad = padarray(img_gray, (2,2), reflect)
img_blur = imfilter(img_pad, Kernel.gaussian(5))
img_thresh = imbinarize.(img_blur[3:end-2, 3:end-2] .|> Mean(), Otsu())

# Find contours
contours = Images.find_contours(img_thresh, 0.5)

# Loop through contours
for cnt in contours
    area = Images.area(cnt)
    if area > 5000
        Images.draw(img, cnt, RGB(0,1,0), 3)
        peri = Images.perimeter(cnt, true)
        approx = Images.approximate_polygon(cnt, 0.02 * peri, keepratio=true)
        if length(approx) == 4
            rect = Images.bounding_rectangle(approx)
            if rect.width > 100 && rect.height > 100
                # Crop and recognize digits using Tesseract.jl
                cell = img_gray[rect.y:rect.y+rect.height-1, rect.x:rect.x+rect.width-1]
                cell_pad = padarray(cell, (2,2), reflect)
                cell_thresh = imbinarize(cell_pad, Otsu())
                text = Tesseract.text(cell_thresh, psm=10)
                if isdigits(text)
                    Images.annotate!(img, text, pos=(rect.x, rect.y), color=RGB(1,0,0), fontsize=20)
                end
            end
        end
    end
end

imshow(img)

File: Sudoku_From_Image/Sudoku_from_Image.py
import cv2
import numpy as np
import pytesseract

# Load image and preprocess
img = cv2.imread('sudoku.jpg')
img_gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
img_blur = cv2.GaussianBlur(img_gray, (5, 5), 1)
img_thresh = cv2.adaptiveThreshold(
    img_blur, 255, cv2.ADAPTIVE_THRESH_GAUSSIAN_C, cv2.THRESH_BINARY, 11, 2)

# Find contours
contours, _ = cv2.findContours(
    img_thresh, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

# Loop through contours
for cnt in contours:
    area = cv2.contourArea(cnt)
    if area > 5000:
        cv2.drawContours(img, cnt, -1, (0, 255, 0), 3)
        peri = cv2.arcLength(cnt, True)
        approx = cv2.approxPolyDP(cnt, 0.02 * peri, True)
        if len(approx) == 4:
            x, y, w, h = cv2.boundingRect(approx)
            if w > 100 and h > 100:
                # Crop and recognize digits using PyTesseract
                cell = img_gray[y:y+h, x:x+w]
                cell_thresh = cv2.adaptiveThreshold(
                    cell, 255, cv2.ADAPTIVE_THRESH_GAUSSIAN_C, cv2.THRESH_BINARY_INV, 11, 2)
                text = pytesseract.image_to_string(
                    cell_thresh, config='--psm 10')
                if text.isdigit():
                    cv2.putText(img, text, (x, y),
                                cv2.FONT_HERSHEY_SIMPLEX, 2, (0, 0, 255), 3)

cv2.imshow('Sudoku', img)
cv2.waitKey(0)
cv2.destroyAllWindows()
